chacha20_poly1305/src/poly1305.rs

```diff
@@ -0,0 +1,242 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Poly1305 one-time message authenticator from RFC8439.
+//!
+//! Heavily inspired by the ["Donna"](https://github.com/floodyberry/poly1305-donna/blob/master/poly1305-donna-32.h) implementation in C
+//! and Loup Vaillant's [Poly1305 design article](https://loup-vaillant.fr/tutorials/poly1305-design).
+
+/// 2^26 for the 26-bit limbs.
+const BITMASK: u32 = 0x03ffffff;
+/// Number is encoded in five 26-bit limbs.
+const CARRY: u32 = 26;
+
+/// Poly1305 authenticator takes a 32-byte one-time key and a message and produces a 16-byte tag.
+///
+/// 64-bit constant time multiplication and addition implementation.
+#[derive(Debug)]
+pub struct Poly1305 {
+    /// r part of the secret key.
+    r: [u32; 5],
+    /// s part of the secret key.
+    s: [u32; 4],
+    /// State used to create tag.
+    acc: [u32; 5],
+    /// Leftovers between adds.
+    leftovers: [u8; 16],
+    /// Track relevant leftover bytes.
+    leftovers_len: usize,
+}
+
+impl Poly1305 {
+    /// Initialize authenticator with a 32-byte one-time secret key.
+    pub fn new(key: [u8; 32]) -> Self {
+        // Taken from Donna. Assigns r to a 26-bit 5-limb number while simultaneously 'clamping' r.
+        let r0 =
+            u32::from_le_bytes(key[0..4].try_into().expect("infalliable conversion")) & 0x3ffffff;
```

Comment:
In 74306fbde17bbbc09e44c388af56a081d8fb7dbf:

This stuff is so ugly that I'd rather just do it manualyly, as

```rust
let r0 = u32::from(key[0]) + (u32::from(key[1]) << 8) + (u32::from(key[2]) << 16) + (u32::from(key[3] << 24);
```

If you use casts instead of `u32::from` you can make the method `const` which is probably worth the uglier code.

Reply:
Ah, I was just doing a deep dive on the`const` with `new` pairing and updating the code based on your comment above and arrived at the same conclusion for this function.

Reply:
I went with bitwise OR because I found it just a tad more readable, but don't have a strong opinion on it. Would be fine with addition.